sonic-xcvrd/xcvrd/xcvrd.py

```diff
@@ -830,6 +849,142 @@ def init_port_sfp_status_tbl(port_mapping, stop_event=threading.Event()):
+            speed = "0"
+            lanes = "0"
+        else:
+            speed = port_change_event.port_dict.get('speed')
+            lanes = port_change_event.port_dict.get('lanes')
+
+        if (pport is None) or (speed is None) or (lanes is None):
+            return
+
+        try:
+            msg = {
+                'lport': lport,
+                'pport': pport,
+                'speed': int(speed),
+                'lanes': lanes.split(',')
+            }
+        except ValueError as ex:
+            self.log_notice("{}".format(ex))
+            return
+
+        self.task_queue.put(msg)
+        time.sleep(0.001)
+
+    def task_worker(self, commander):
+        self.log_notice("Starting...")
+
+        if commander:
+            sel, asic_context = port_mapping.subscribe_port_config_change()
+            while not self.task_stopping_event.is_set():
+                # Handle port change event from main thread
+                port_mapping.handle_port_config_change(sel, asic_context, self.task_stopping_event, self.port_mapping, helper_logger, self.on_port_config_change)
+        else:
+            while not self.task_stopping_event.is_set():
+                try:
+                    msg = self.task_queue.get(False, 0.1)
+                except:
+                    msg = None
+                if msg is None:
+                    time.sleep(1)
+                    continue
+
+                lport = msg['lport']
+                pport = msg['pport']
+                lanes = msg['lanes']
+                speed = msg['speed']
+                if speed == 0 or len(lanes) < 1:
+                    continue
+                sfp = platform_chassis.get_sfp(pport)
+                if not sfp.get_presence():
+                    continue
+
+                try:
+                    # Skip if the port/cage type is not QSFP-DD
+                    if sfp.SFP_PORT_TYPE_QSFPDD != sfp.get_port_type():
+                        continue
+                    # Skip if the xcvr type is not QSFP-DD
+                    info = sfp.get_transceiver_info()
+                    if info.get('type_abbrv_name', None) not in ['QSFP-DD', 'QSFP_DD']:
+                        continue
+                    # Skip if the memory type is flat
+                    if info.get('memory_type', 'flat') in ['flat', 'FLAT']:
+                        continue
+
+                    # Fetch the custom CMIS media setting from media_settings.json
+                    conf = get_media_settings_cmis(info)
+                    if conf is None:
+                        conf = {}
+                    # Enable application selection by default
+                    if conf.get('application_selection', True):
+                        self.log_notice("{}: application update for {}G, {}-lanes".format(
+                            lport, int(speed/1000), len(lanes)))
+                        ret = sfp.set_cmis_application(speed, len(lanes))
```

Comment:
How do we ensure that this is done _after_ the ASIC SerDes taps are configured?

Reply:
If you're referring to the pre-emphasis configuration on the MAC, it's handled by the orchagent when the serdes tap parameters are uploaded to the APPL_DB, hence it's outside the scope of this PR
If you're referring to the pre-emphasis configuration on the PHY, it's not supported as of now, we could circle back on this when the module-specific parameters are available and needs to be applied to the CMIS PHY registers.

Reply:
Note: In the case of Broadocm ASIC, we have different value range for PAM4 and NRZ, hence having NRZ/PAM4 pre-emphasise applied to SerDes in PAM4/NRZ mode will lead to checker failure in SAI, and in turn, the orchagent will crash.

Reply:
My concern is that (per my understanding) the ASIC SerDes pre/main/post setitngs should be configured before the CMIS ApSel sequence is started. There doesn't seem to be anything that enforces this sequencing here.

Reply:
Yes, we need a synchronization mechanism between syncd/SAI and the xcvrd to gracefully switch in between the applications. Prince @ MSFT and I had this discussion yesterday, and he will be working on this, and we could deal with this in another PR